Enable opening of dummy data

xulman opened the pull request:
Previously there was BasicImgLoader utlized in `DummySpimData` object.
But [some code around](https://github.com/mastodon-sc/mastodon/blob/adc7bddb4850510007b1a93004f33d04b7bb2340/src/main/java/org/mastodon/views/bdv/SharedBigDataViewerData.java#L392) (even [outside Mastodon](https://github.com/bigdataviewer/bigdataviewer-core/blob/9255efe8eb63a4b861ea7c6be283b556d31c0e7a/src/main/java/bdv/BigDataViewer.java#L277)) is casting to `ViewerImgLoader`.

To enable these pieces of code to function with our dummy data, we had to upgrade the dummy....

The implementation features single resolution level pyramid (I hope).

xulman commented on src/main/java/org/mastodon/util/DummySpimData.java:
```diff
@@ -139,22 +144,65 @@ private static double get( final String[] parts, final String key, final double
+				final double[][] resolution = new double[1][3];
+				final AffineTransform3D[] transform = new AffineTransform3D[1];
+				{
+					resolution[0][0] = 1.0; //TODO should be taken from 'calib' from our only-one caller code...
```
I'm not sure how much important is to fix this TODO given that we talk here about DUMMY data...

if we were to make it crystal clear, we would need to modify the c'tor (easy, there is only one caller of it) to bring in ["calibration" and "transformation"](https://github.com/mastodon-sc/mastodon/blob/adc7bddb4850510007b1a93004f33d04b7bb2340/src/main/java/org/mastodon/util/DummySpimData.java#L110-L120)

tinevez commented:
Maybe this PR could be the opportunity to improve opening a mastodon file when the image data is missing. For now several things could be improved:

- [ ] The user gets a convoluted error message in the IJ console. We could give them a nice warning window, likethe ones we use when opening from ImageJ ImagePlus.
![image](https://user-images.githubusercontent.com/3583203/190895305-2de1a7f1-4293-4325-8f33-5ab9f81fbfcb.png)

- [ ] The BDV appears empty of tracks when you open it. The user must dezoom or pan or zoom to make them appear. Maybe because of an incorrect bounding box determination at start?
![image](https://user-images.githubusercontent.com/3583203/190895391-cb6d95d7-7187-43b8-a241-755078247880.png)

- [ ] Same thing for time: TrackScheme opens with a max timepoint of 9 and cannot zoom out. It does not determine max timepoint from the data:
![image](https://user-images.githubusercontent.com/3583203/190895426-fd7a3adb-31e2-4fa8-a74a-1ffb5079cc1e.png)

xulman commented:
yes, agree to all

btw, I've revamped the File -> Fix Image Path dialog in mastodon-tomancak repo recently;
on the master in m.-tomancak repo it looks now like this:  
![Screenshot_20220919_115401](https://user-images.githubusercontent.com/10509335/190993304-75786dc1-e8ab-405a-9c6e-30568104a75d.png)

In the dummy sub dialog I could maybe pre-load default values from the `Graph`, and we could try to open this thing whenever loading of project fails?

what do you think? @tinevez @tpietzsch @maarzt

xulman commented:
so, to the latter two points of JY above:
the dummy data is some default `Interval`, likely small.. and thus off compated to where the spots actually are, thus the BDV view looks at nothing

similarly to the timpoints range, the dummy data by default creates only 10 timepoint, which is why TS win was "trimmed" (also BDV window should have time slider only for 10 TPs)

while i didn't care too much about the spatial extent of the dummy data (it's empty, why would I even cared to open BDV...), I cared about the temporal limit and that actually motivated me to introduce the "I want dummy data" button in which *especially* the time span can be adjusted

xulman commented:
actually was now tracing how default dummy data is created:
- https://github.com/mastodon-sc/mastodon/blob/9b786081652a2975bf84948281cdc2db8bf75b79/src/main/java/org/mastodon/mamut/ProjectManager.java#L600
- [if we have some `.xml` data pointer, we open it directly here](https://github.com/mastodon-sc/mastodon/blob/9b786081652a2975bf84948281cdc2db8bf75b79/src/main/java/org/mastodon/mamut/ProjectManager.java#L639)
- [it then assumes it is a BDV data](https://github.com/mastodon-sc/mastodon/blob/9b786081652a2975bf84948281cdc2db8bf75b79/src/main/java/org/mastodon/mamut/ProjectManager.java#L925)
- [after failing opening the data, it inspects...](https://github.com/mastodon-sc/mastodon/blob/9b786081652a2975bf84948281cdc2db8bf75b79/src/main/java/org/mastodon/views/bdv/SharedBigDataViewerData.java#L384)
- [...with defaults](https://github.com/mastodon-sc/mastodon/blob/9b786081652a2975bf84948281cdc2db8bf75b79/src/main/java/org/mastodon/util/DatasetInfoParser.java#L104) which is where the 10 TPs are set

hmm.. it's a puzzle... when the `.xml` file is missing, it does not even reach the `inspect()` code...

xulman commented:
* rebased against the latest `origin/dev` brach
* added one commit that attempts to clean up a bit the console reporting... the main contribution is basically only the replacing of `printStackTrace()` with `getMessage()`... no big, but looks less messy

xulman commented:
so, this is what it looks like:
![Screenshot_20220919_163028](https://user-images.githubusercontent.com/10509335/191042093-2d4e218e-ed55-4707-a78c-7091c1b43d55.png)

now, since it actually opens something, two consequences:
- needs to open something to be able to fix its dataset path
- the launcher window is gone and makes no sense putting info/warn/error msg into it

@tinevez I would say this could close the item 1 "user gets convoluted message", what do you think?

xulman commented:
![Screenshot_20220919_171347](https://user-images.githubusercontent.com/10509335/191054036-8323ded7-e13b-4b5e-a15f-5a74811a1771.png)

@tinevez I would say this could close the item 1 "user gets convoluted message", what do you think?

xulman commented:
The fix dialog path, in its "dummy button", got option to adjust the dummy data to be just "around the spots" (aka axis aligned bounding box in spatial axes and temporal axis)... 
![Screenshot_20220919_190545](https://user-images.githubusercontent.com/10509335/191073872-b123f241-3e0f-4cd9-9701-61e0c13488c4.png)

you can find it in [tomancak mastodon repo, branch aabb-dummy](https://github.com/mastodon-sc/mastodon-tomancak/tree/aabb-dummy)

tinevez commented:
Hello @xulman 
I think this PR is ready to be merged. 
Can you briefly tell me of a procedure that would allow me to test it?

xulman commented:
- take the latest from https://github.com/xulman/mastodon/tree/enableOpeningOfDummyData (which is a fork of the official main mastodon/dev branch on which 4 commits to handle the dummy data are added)
- take the latest commit from [mastodon-tomacak repo](https://github.com/mastodon-sc/mastodon-tomancak), master branch
- compile & run
- take any mastodon project, copy it and cripple the path to the BDV data in the copy (e.g. introduce typo in the path in the project.xml or just move the data itself to another folder)
- try to open this project...
